### PR TITLE
Visualize retaining path.

### PR DESCRIPTION
**Pull request by polina-c**

### CODE REVIEW

**Comment by polina-c** [comment]
https://user-images.githubusercontent.com/12115586/196534926-f1ce02a4-2405-4b98-8ef3-40c51057df5c.mov

File: packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/class_details/class_details.dart [hunk 1]
```diff
@@ -28,10 +35,35 @@ class HeapClassDetails extends StatelessWidget {
...
+          path: selection.key,
+          controller: pathController,
+        );
+      },
+    );
+
+    return Split(
+      axis: Axis.horizontal,
+      initialFractions: const [0.7, 0.3],
+      children: [area1, area2],
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/class_details/class_details.dart:66]: Nit: rename these variables to something more descriptive. Maybe `retainingPathView` and `statsView`?

File: packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/class_details/path.dart [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      children: [
+        const SizedBox(height: densePadding),
+        _PathControlPane(
+          controller: controller,
+          path: path,
+        ),
+        Expanded(
+          child: Padding(
+            padding:
+                const EdgeInsets.only(top: densePadding, left: densePadding),
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/class_details/path.dart:35]: Nit: trailing comma after `densePadding`.

File: packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart [hunk 3]
```diff
@@ -126,10 +126,69 @@ class ClassOnlyHeapPath {
             heapPath.objects.map((o) => o.heapClass).toList(growable: false);
   final List<HeapClassName> classes;
 
-  String asShortString() => classes.map((e) => e.className).join('/');
+  String asShortString({String? delimiter, bool inverted = false}) => _asString(
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart:129]: Nit: consider using `toShortString()` and `toLongString()` instead.

File: packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart [hunk 4]
```diff
@@ -126,10 +126,69 @@ class ClassOnlyHeapPath {
...
+        ),
+        inverted: inverted,
+      );
+
+  String asLongString({
+    String? delimiter,
+    bool inverted = false,
+    bool hideStandard = false,
+  }) {
+    late List<String> data;
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart:144]: Does this need to be late? I would expect static analysis to have noticed that this is initialized unconditionally below.

File: packages/devtools_app/lib/src/screens/memory/panes/diff/controller/simple_controllers.dart [hunk 5]
```diff
@@ -0,0 +1,10 @@
+// Copyright 2022 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+class RetainingPathController {
+  final hideStandard = ValueNotifier<bool>(true);
```
Comment (bkonyi) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/controller/simple_controllers.dart:8]: Nit: I generally try and avoid exposing `ValueNotifier`s directly. If these are just meant to be toggled, maybe use `ValueListenable` getters and add `toggleHideStandard()` and `toggleInvert()` methods?

**Comment by bkonyi** [comment]
> Screen.Recording.2022-10-18.at.1.15.10.PM.mov

Thanks for the video! I do still think that the set of retaining paths for an instance might be better visualized as a tree, but I haven't had a chance to play around with your implementation quite yet.

File: packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart (hunk 4 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart:144]: wow! indeed
thanks, i did not expect it to be possible :)

**Comment by polina-c** [comment]
> > Screen.Recording.2022-10-18.at.1.15.10.PM.mov
> 
> Thanks for the video! I do still think that the set of retaining paths for an instance might be better visualized as a tree, but I haven't had a chance to play around with your implementation quite yet.

You can add this suggestion to go/ux-improvements and link a doc or issue with details/screenshots. I hope to make the process around the doc alive at some moment.

**Pull request merged**